gson/src/org/immutables/gson/stream/JsonGeneratorWriter.java

```diff
@@ -145,7 +148,23 @@ public JsonWriter value(Number value) throws IOException {
         throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
       }
     }
-    generator.writeNumber(d);
+    if (value instanceof Integer) {
+      generator.writeNumber(value.intValue());
+    } else if (value instanceof Short) {
+      generator.writeNumber(value.shortValue());
+    } else if (value instanceof Long) {
+      generator.writeNumber(value.longValue());
+    } else if (value instanceof Float) {
+      generator.writeNumber(value.floatValue());
+    } else if (value instanceof Double) {
+      generator.writeNumber(value.doubleValue());
+    } else if (value instanceof BigInteger) {
+      generator.writeNumber((BigInteger) value);
+    } else if (value instanceof BigDecimal) {
+      generator.writeNumber((BigDecimal) value);
+    } else {
+      generator.writeNumber(d);
```

Comment:
why not `generator.writeNumber(value.doubleValue())` ?

Reply:
Fine but in it will mean in cases of a double value with strict checks, doubleValue will Be called twice

Reply:
you're talking about `isLinient()`

```java
} else {
   // remove previous "value instanceof Double"
    double d = value.doubleValue();
    if (!isLenient()) {
      if (Double.isNaN(d) || Double.isInfinite(d)) {
        throw new IllegalArgumentException("JSON forbids NaN and infinities: " + value);
      }
    }
   generator.writeNumber(d);
}
```